transports/quic/src/transport.rs

```diff
@@ -223,14 +295,130 @@ impl<P: Provider> Transport for GenTransport<P> {
+                    } else {
+                        return Poll::Ready(TransportEvent::Incoming {
+                            listener_id,
+                            upgrade,
+                            local_addr,
+                            send_back_addr,
+                        });
+                    }
+                }
+                _ => return Poll::Ready(ev),
+            }
         }
 
         self.waker = Some(cx.waker().clone());
         Poll::Pending
     }
 }
 
+struct HolePuncher {
+    endpoint_channel: endpoint::Channel,
+    remote_addr: SocketAddr,
+    timeout: Delay,
+    interval_timeout: Delay,
+    receiver: Receiver<Connecting>,
+    connecting: Option<Connecting>,
+}
+
+impl HolePuncher {
+    fn new(
+        endpoint_channel: endpoint::Channel,
+        remote_addr: SocketAddr,
+        timeout: Duration,
+        receiver: Receiver<Connecting>,
+    ) -> Self {
+        Self {
+            endpoint_channel,
+            remote_addr,
+            timeout: Delay::new(timeout),
+            interval_timeout: Delay::new(Duration::from_secs(0)),
+            receiver,
+            connecting: None,
+        }
+    }
+}
+
+impl Future for HolePuncher {
+    type Output = Result<(PeerId, Connection), Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if let Some(connecting) = &mut self.connecting {
+            match connecting.poll_unpin(cx) {
+                Poll::Pending => {}
+                ready => return ready,
+            }
+        }
+
+        match self.receiver.poll_unpin(cx) {
+            Poll::Ready(Ok(upgrade)) => {
+                self.connecting = Some(upgrade);
+            }
+            Poll::Ready(Err(_)) => {}
+            Poll::Pending => {}
+        }
+
+        match self.timeout.poll_unpin(cx) {
+            Poll::Ready(_) => return Poll::Ready(Err(Error::HandshakeTimedOut)),
+            Poll::Pending => {}
+        }
+
+        match self.interval_timeout.poll_unpin(cx) {
+            Poll::Ready(_) => {
+                let message = ToEndpoint::SendUdpPacket(quinn_proto::Transmit {
+                    destination: self.remote_addr,
+                    ecn: None,
+                    contents: rand::thread_rng()
+                        .sample_iter(distributions::Standard)
+                        .take(64)
+                        .collect(),
+                    segment_size: None,
+                    src_ip: None,
+                });
+
+                match self.endpoint_channel.try_send(message, cx) {
+                    Ok(_) => {}
+                    Err(endpoint::Disconnected {}) => {
+                        return Poll::Ready(Err(Error::EndpointDriverCrashed))
+                    }
+                }
+
+                self.interval_timeout.reset(Duration::from_millis(
+                    rand::thread_rng().gen_range(10..=200),
+                ));
+            }
+            Poll::Pending => {}
+        }
```

Comment:
Would it make sense to model this as a separate task that gets spawned? We could combine it with a [`RemoteHandle`](https://docs.rs/futures/latest/futures/future/struct.RemoteHandle.html) and store it in the `ActiveHolePunch` struct. As soon as we deallocate that one, the task would be stopped.